### ISSUE TITLE
Singularity fails when large environment is sourced

### ISSUE THREAD

**Issue by kt-pham** [apptainer/singularity#5946]
### Version of Singularity:

```
> singularity --version
singularity version 3.7.3-1

```

### Expected behavior

```
> repeat 100 singularity exec sles12sp5.sif uptime
16:22pm  up 58 days 16:19,  0 users,  load average: 49.79, 49.44, 49.30
16:22pm  up 58 days 16:19,  0 users,  load average: 50.92, 49.71, 49.39
```

### Actual behavior
```
> repeat 100 singularity exec sles12sp5.sif uptime

runtime: netpoll: break fd ready for 16
fatal error: runtime: netpoll: break fd ready for something unexpected

runtime stack:
runtime.throw(0x555c126b9e27, 0x39)
        runtime/panic.go:1116 +0x74
runtime.netpoll(0x0, 0x0)
        runtime/netpoll_epoll.go:149 +0x2ee
runtime.sysmon()
        runtime/proc.go:4714 +0x34f
runtime.mstart1()
        runtime/proc.go:1172 +0xca
runtime.mstart()
        runtime/proc.go:1137 +0x6f

goroutine 1 [runnable, locked to thread]:
github.com/sylabs/singularity/internal/pkg/util/shell/interpreter.GetEnv.func1(0xc000179950, 0x2d, 0x1000100, 0xc00017997e, 0x60, 0x0, 0x0, 0x0, 0x0, 0x1)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/shell/interpreter/interpreter.go:93 +0x50
mvdan.cc/sh/v3/expand.listEnviron.Each(0xc00038a000, 0xaac, 0xc00, 0xc000228970)
        mvdan.cc/sh/v3@v3.2.1/expand/environ.go:206 +0x15d
mvdan.cc/sh/v3/interp.overlayEnviron.Each(0x555c12a7a6c0, 0xc000322180, 0xc000133920, 0xc000228970)
        mvdan.cc/sh/v3@v3.2.1/interp/vars.go:33 +0x5b
github.com/sylabs/singularity/internal/pkg/util/shell/interpreter.GetEnv(0x555c12a7a740, 0xc000322800, 0xc00019ab40, 0x14, 0x555c12a73e60, 0xc0001b0000, 0x555c12a72ec0, 0xc0001eb430, 0x555c12a73ea0, 0xc0001b0010, ...)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/shell/interpreter/interpreter.go:92 +0x96
github.com/sylabs/singularity/internal/pkg/runtime/engine/singularity.getAllEnvBuiltin.func1(0x555c12a80fc0, 0xc000133950, 0xc000228920, 0x0, 0x0, 0x1, 0x0)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/runtime/engine/singularity/process_linux.go:716 +0x17c
github.com/sylabs/singularity/internal/pkg/util/shell/interpreter.(*Shell).internalExecHandler.func1(0x555c12a80fc0, 0xc000133950, 0xc000228920, 0x1, 0x1, 0xc000228920, 0x555c1264d21e)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/shell/interpreter/interpreter.go:153 +0x37f
mvdan.cc/sh/v3/interp.(*Runner).exec(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0xc000228920, 0x1, 0x1)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:753 +0x8a
mvdan.cc/sh/v3/interp.(*Runner).call(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0x1f000300000029, 0xc000228920, 0x1, 0x1)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:749 +0x405
mvdan.cc/sh/v3/interp.(*Runner).cmd(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0x555c12a7edc0, 0xc000396a80)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:345 +0x14c8
mvdan.cc/sh/v3/interp.(*Runner).stmtSync(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0xc000398058)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:285 +0x312
mvdan.cc/sh/v3/interp.(*Runner).stmt(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0xc000398058)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:266 +0x1e7
mvdan.cc/sh/v3/interp.(*Runner).stmts(0xc000110400, 0x555c12a80f40, 0xc00019c008, 0xc00039e000, 0x1, 0x4)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:603 +0x58
mvdan.cc/sh/v3/interp.(*Runner).fillExpandConfig.func1(0x555c12a72ec0, 0xc0001eb430, 0xc0001805a0, 0xc00020d0c8, 0x555c1264c689)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:49 +0xe6
mvdan.cc/sh/v3/expand.(*Config).cmdSubst(0xc0001eb400, 0xc0001805a0, 0x3, 0x0, 0x0, 0x0)
        mvdan.cc/sh/v3@v3.2.1/expand/expand.go:500 +0x70
mvdan.cc/sh/v3/expand.(*Config).wordField(0xc0001eb400, 0xc0000ab420, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        mvdan.cc/sh/v3@v3.2.1/expand/expand.go:471 +0x9d3
mvdan.cc/sh/v3/expand.Literal(0xc0001eb400, 0xc000224c30, 0x0, 0x0, 0x0, 0x0)
        mvdan.cc/sh/v3@v3.2.1/expand/expand.go:143 +0x7a
mvdan.cc/sh/v3/interp.(*Runner).literal(0xc000110200, 0xc000224c30, 0x10, 0x0)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:155 +0x45
mvdan.cc/sh/v3/interp.(*Runner).assignVal(0xc000110200, 0xc0001323c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        mvdan.cc/sh/v3@v3.2.1/interp/vars.go:263 +0x955
mvdan.cc/sh/v3/interp.(*Runner).cmd(0xc000110200, 0x555c12a80f40, 0xc00019c008, 0x555c12a7ef00, 0xc000322200)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:517 +0x1fd6
mvdan.cc/sh/v3/interp.(*Runner).stmtSync(0xc000110200, 0x555c12a80f40, 0xc00019c008, 0xc000398000)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:285 +0x312
mvdan.cc/sh/v3/interp.(*Runner).stmt(0xc000110200, 0x555c12a80f40, 0xc00019c008, 0xc000398000)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:266 +0x1e7
mvdan.cc/sh/v3/interp.(*Runner).stmts(0xc000110200, 0x555c12a80f40, 0xc00019c008, 0xc0001fe400, 0x16, 0x20)
        mvdan.cc/sh/v3@v3.2.1/interp/runner.go:603 +0x58
mvdan.cc/sh/v3/interp.(*Runner).Run(0xc000110200, 0x555c12a80f40, 0xc00019c008, 0x555c12a79a40, 0xc0000ce840, 0xc0000ce840, 0x0)
        mvdan.cc/sh/v3@v3.2.1/interp/api.go:504 +0x2ea
github.com/sylabs/singularity/internal/pkg/util/shell/interpreter.(*Shell).Run(0xc0000ce800, 0xc000132360, 0x555c12691155)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/shell/interpreter/interpreter.go:223 +0x145
github.com/sylabs/singularity/internal/pkg/runtime/engine/singularity.runActionScript(0xc0001a9220, 0x17, 0x0, 0x0, 0x12, 0x0, 0x7f976e79f548, 0x8, 0x80)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/runtime/engine/singularity/process_linux.go:831 +0x775
github.com/sylabs/singularity/internal/pkg/runtime/engine/singularity.(*EngineOperations).StartProcess(0xc000228860, 0x555c12a85c20, 0xc000010c50, 0x0, 0x0)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/runtime/engine/singularity/process_linux.go:171 +0x5ac
github.com/sylabs/singularity/internal/app/starter.StageTwo(0x4, 0xc00000e0c0)
        github.com/sylabs/singularity@v0.0.0/internal/app/starter/stage_linux.go:58 +0x15b
main.startup.func1()
        command-line-arguments/main_linux.go:64 +0x3c
github.com/sylabs/singularity/internal/pkg/util/mainthread.Execute.func1()
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/mainthread/mainthread.go:21 +0x31
main.main()
        command-line-arguments/main_linux.go:106 +0x62

goroutine 5 [chan receive]:
github.com/sylabs/singularity/internal/pkg/util/mainthread.Execute(0xc000021e20)
        github.com/sylabs/singularity@v0.0.0/internal/pkg/util/mainthread/mainthread.go:24 +0xb6
main.startup()
        command-line-arguments/main_linux.go:63 +0x24f
created by main.main
        command-line-arguments/main_linux.go:102 +0x37

goroutine 6 [runnable]:
os/signal.loop()
        os/signal/signal_unix.go:21
created by os/signal.Notify.func1.1
        os/signal/signal.go:150 +0x46
```

### Steps to reproduce this behavior

Our environment has 2727 entries using 629,101 bytes. When the environment is sourced and singularity is executed, the above error is seen, but not consistently. Out of 100 executions, up to 10% see the issue only on machines with a higher load average. 

It seems that we are only able to reproduce this issue with both the large environment _and_ on a machine with a higher load average (load > cores). 

### What OS/distro are you running

We have tested and seen this error on SLES12-SP2, SLES12-SP5 and SLES15-SP2. 

### How did you install Singularity

rpmbuild -tb singularity-3.7.3.tar.gz
rpm -ivh /usr/src/packages/RPMS/x86_64/singularity-3.7.3-1.x86_64.rpm

**Comment by dtrudg** [comment]
I believe this is the same issue as something similar and sporadic that's been reported to us elsewhere. I'll write it up shortly here when that's confirmed.

**Comment by xspython** [comment]
thx!

**Comment by mfarooq-tech** [comment]
@dtrudg : Any updates?

**Comment by dtrudg** [comment]
I haven't had time to setup your reproducer yet to confirm. It's on my list.

**Comment by dtrudg** [comment]
This is a tricky one. We've had a separate report of what appears to be the same issue, triggered very infrequently (1 in a million executions scale) with a slightly different backtrace (it likely varies based on version of Go used to build Singularity). Thanks for the detail here that gives a possible reproducer!

Go is a language with a runtime that includes a parallelized garbage collector. The netpoll failure on the fd is related to a GC memory profiling operation.

When Singularity is advanced in the container setup process it must close any file descriptors that are not associated with the container, or the container setup process. The container environment is sourced by a Go embedded shell interpreter. While the Go runtime could trigger a GC / memory profiling cycle at various times, a large environment is more likely to trigger a GC cycle / GC memory profiling operation in this step. 

There appears to be a conflict between Singularity identifying and closing file descriptors, and Go runtime GC operations that may occur. The Go GC cycle isn't completely 'halt the world', perhaps explaining why high load and large environment are required to trigger it, as they possibly delay the completion of the starter execution path as some of the GC stuff happens in parallel.

The following change has been observed to work around this, by disabling Go runtime GC for the short-lived starter process that closes fds and performs other final container setup:

Allow the starter to keep variables that allow us to tweak the Go runtime GC behavior for the starter process:

```diff
diff --git a/cmd/starter/c/starter.c b/cmd/starter/c/starter.c
index a99de873b..678a32e7f 100644
--- a/cmd/starter/c/starter.c
+++ b/cmd/starter/c/starter.c
@@ -975,6 +975,7 @@ static void cleanup_fd(fdlist_t *master, struct starter *starter) {
             if ( starter->fds[i] == fd ) {
                 found = true;
                 /* set force close on exec */
+                debugf("Setting FD_CLOEXEC on starter fd %d\n", starter->fds[i]);
                 if ( fcntl(starter->fds[i], F_SETFD, FD_CLOEXEC) < 0 ) {
                     debugf("Can't set FD_CLOEXEC on file descriptor %d: %s\n", starter->fds[i], strerror(errno));
                 }
@@ -1128,9 +1129,17 @@ static void cleanenv(void) {
     /*
      * keep only SINGULARITY_MESSAGELEVEL for GO runtime, set others to empty
      * string and not NULL (see issue #3703 for why)
+     *
+     * DCT - also keep any GOGC and GODEBUG vars for go runtime
+     * debugging purposes.
      */
     for (e = environ; *e != NULL; e++) {
-        if ( strncmp(MSGLVL_ENV "=", *e, sizeof(MSGLVL_ENV)) != 0 ) {
+        if ( strncmp(MSGLVL_ENV "=", *e, sizeof(MSGLVL_ENV)) == 0 ||
+             strncmp("GOGC" "=", *e, sizeof("GOGC")) == 0 ||
+             strncmp("GODEBUG" "=", *e, sizeof("GODEBUG")) == 0 ) {
+            debugf("Keeping env var %s\n", *e);
+        } else {
+            debugf("Clearing env var %s\n", *e);
             *e = "";
         }
     }
```

Disable garbage collections altogether for starter by setting `GOGC=off` (this is the fix/workaround)

Note the `GODEBUG=` line allows finer grained debugging. In this example it's turning off GC mem profiling and instructing Go to print a trace of any GC operations. If the `GOGC=off` is present this shouldn't do anything.

```diff
diff --git a/internal/pkg/util/starter/starter.go b/internal/pkg/util/starter/starter.go
index e87abead3..92cca08cf 100644
--- a/internal/pkg/util/starter/starter.go
+++ b/internal/pkg/util/starter/starter.go
@@ -89,6 +89,10 @@ func Exec(name string, config *config.Common, ops ...CommandOp) error {
 	if err := c.init(config, ops...); err != nil {
 		return fmt.Errorf("while initializing starter command: %s", err)
 	}
+	sylog.Debugf("Setting GOGC=off for starter")
+	c.env = append(c.env, "GOGC=off")
+	sylog.Debugf("Setting GODEBUG=memprofilerate=0,gctrace=1 for starter")
+	c.env = append(c.env, "GODEBUG=memprofilerate=0,gctrace=1")
 	err := unix.Exec(c.path, []string{name}, c.env)
 	return fmt.Errorf("while executing %s: %s", c.path, err)
 }
```

@mohafarooq - would you able to apply this and confirm if it does or doesn't resolve the failure you are seeing? I believe it will... but I could easily be wrong :-)

**Comment by kt-pham** [comment]
@dtrudg - Is there a way to verify GOGC=off after the singularity RPM is built?

**Comment by dtrudg** [comment]
> @dtrudg - Is there a way to verify GOGC=off after the singularity RPM is built?

Check for the message the code here generates when running in debug mode `singularity -d`

```
sylog.Debugf("Setting GOGC=off for starter")
```

**Comment by kt-pham** [comment]
Thanks @dtrudg! Using the fix above, I see only 1 failure out of nearly 50,000 executions (using the same parameters as above). 

Will you be able to provide this fix as part of the next release?

**Comment by carterpeel** [comment]
Hello,

This is a templated response that is being sent out to all open issues.  We are working hard on 'rebuilding' the Singularity community, and a major task on the agenda is finding out what issues are still outstanding.


**Please consider the following:**

1. Is this issue a duplicate, or has it been fixed/implemented since being added?
2. Is the issue still relevant to the current state of Singularity's functionality?
3. Would you like to continue discussing this issue or feature request?


Thanks,
Carter

**Comment by kt-pham** [comment]
@carterpeel @dtrudg 

Is this issue a duplicate, or has it been fixed/implemented since being added? --> No
Is the issue still relevant to the current state of Singularity's functionality? --> Yes
Would you like to continue discussing this issue or feature request? --> Yes

**Comment by rdenham** [comment]
Just wanted to chime in and report that this affects users on our system as well, although we typically don't get any output at all, the command just fails to do anything. I had also narrowed it down to the size of the users environment. If you use `--clean-env` then there is no problem. If you reduce the size of the environment also no problem. I always just assumed that our ridiculously large environment was unique to us so this was something that didn't affect anybody else. I have taken steps to reduce the size of users' environments to keep singularity working as expected.

**Comment by mfarooq-tech** [comment]
@dtrudg : It will be great if you can commit the suggested code and let us know the plans to release the same.

**Comment by dtrudg** [comment]
Hi @mohafarooq - Apologies. I no longer work on this repository, following the fork that was announced on the mailing list. 

I understand that @kmuriki has taken over as the release manager for hpcng/singularity

Apologies for the confusion.

**Comment by gmkurtzer** [comment]
It looks like Sylabs put a fix into their own repo almost immediately after they forked the project. Typically forks push/offer fixes back to the source open source community, but apparently and unfortunately, that isn't happening here.

@kmuriki I did a quick check, and I found this in the Sylabs commercial git:

https://github.com/sylabs/singularity/commit/1137d8be666fc35fce00cc73bba06e60f0a3b86d

We should either cherry-pick it, or propose a different fix as this didn't seem to be a 100% solution. @cclerget and/or @ikaneshiro might be able to shed some additional light on that.

Greg

**Comment by dtrudg** [comment]
Please note that the sylabs/singularity (SingularityCE) fork is an open-source project, and remains under the open source BSD license. It is not a commercial variant of Singularity.

Code merged to the repository is publicly visible, and anyone is welcome to use it under the terms of the license. All fixes / features we make are available/offered on this basis.

**Comment by DrDaveD** [comment]
Sylabs did this in their [pr 8](https://github.com/sylabs/singularity/pull/8) and put it in their 3.8.0 but @cclerget would like to do this in a different way.

**Comment by cclerget** [comment]
I wasn't able to reproduce on bare ubuntu 20.04 and centos 8 VM, on ubuntu 20.04 I checked with both go 1.16 and 1.15. But here the procedure to attempt to reproduce it:

In the first terminal:
```
a=$(python3 -c "print('A'*500)"
for i in $(seq 1 2000); do export A$i="$a"; done
```

The above results in an environment size of `1016467`.

Simulate overload on the machine from the second terminal:
```
stress-ng -c 96 --vm-bytes=32G --vm 4
```

And from the first terminal execute:
```
for i in $(seq 1 100); do singularity exec alpine.sif id; done
```

I also checked that closed file descriptors are not part of the netpoll fds with this patch
and the message `NETPOLL FD` never show up:

```
diff --git a/internal/pkg/runtime/engine/singularity/process_linux.go b/internal/pkg/runtime/engine/singularity/process_linux.go
index 73272b4d3..c4fc82422 100644
--- a/internal/pkg/runtime/engine/singularity/process_linux.go
+++ b/internal/pkg/runtime/engine/singularity/process_linux.go
@@ -47,6 +47,9 @@ import (
        "mvdan.cc/sh/v3/interp"
 )
 
+//go:linkname isPollServerDescriptor internal/poll.runtime_isPollServerDescriptor
+func isPollServerDescriptor(fd uintptr) bool
+
 const defaultShell = "/bin/sh"
 
 // StartProcess is called during stage2 after RPC server finished
@@ -129,12 +132,20 @@ func (e *EngineOperations) StartProcess(masterConn net.Conn) error {
                // partition is considered as a single image with
                // different offset/size but pointing to the same
                // opened image file descriptor
+               if isPollServerDescriptor(uintptr(img.Fd)) {
+                       fmt.Println("NETPOLL FD")
+                       continue
+               }
                if err := syscall.Close(int(img.Fd)); err != nil && err != syscall.EBADF {
                        return fmt.Errorf("failed to close file descriptor for %s: %s", img.Path, err)
                }
        }
 
        for _, fd := range e.EngineConfig.GetOpenFd() {
+               if isPollServerDescriptor(uintptr(fd)) {
+                       fmt.Println("NETPOLL FD")
+                       continue
+               }
                if err := syscall.Close(fd); err != nil {
                        return fmt.Errorf("aborting failed to close file descriptor: %s", err)
                }
```

Additionally it looks like the Sylabs patch greatly reduce the number of failures but doesn't fix it completely as stated by @kt-pham (https://github.com/hpcng/singularity/issues/5946#issuecomment-832293360)  which indicates this may not be the right fix.

So the issue seems more related to the masterConn socket here https://github.com/hpcng/singularity/blob/eb85a9a928a2368a610004d4ad382be9061a452b/internal/pkg/runtime/engine/singularity/process_linux.go#L58
staying open until the container process is executed or closed explicitly, this is a requirement to communicate with the parent process and can only be closed after the execution of the container process.

What happen here is that the stage2 common code create a `net.FileConn` from the master socket descriptor here https://github.com/hpcng/singularity/blob/eb85a9a928a2368a610004d4ad382be9061a452b/internal/app/starter/stage_linux.go#L46
therefore the file descriptor become part of the netpoll fds.

So the fix (#6031) would be to pass this file descriptor under the Go runtime radar by not using `net.FileConn` and dealing with it at a lowest level directly with syscalls.

Would be much appreciated if you guys @kt-pham @rdenham @mohafarooq could give it a try and feedbacks. Thanks !
